### PR TITLE
Document access_tokens and refresh_tokens

### PR DESCRIPTION
**Pull request by richvdh**
Hopefully, describe the rationale and mechanisms of access tokens and refresh tokens.

Fixes https://matrix.org/jira/browse/SPEC-343

### CODE REVIEW

File: specification/client_server_api.rst [hunk 1]
```diff
@@ -158,14 +158,109 @@ recommended.
...
+       |        |                                       |        |
+       |        | <-(D)--------- Response ------------- |        |
+       |        |                                       |        |
+       |        |                                       |        |
+       |        | --(E)-- Request with access_token --> |        |
+       | Client |                                       | Server |
+       |        | <-(F)------ M_UNKNOWN_TOKEN --------- |        |
+       |        |                                       |        |
+       |        |                                       |        |
+       |        | --(G)------ Token refresh ---- -----> |        |
```
Comment (ara4n) [inline specification/client_server_api.rst]: shouldn't this be "Token refresh, with refresh token"?

File: specification/client_server_api.rst [hunk 2]
```diff
@@ -158,14 +158,109 @@ recommended.
...
 ``M_UNKNOWN_TOKEN`` respectively.
 
+Access tokens may have a limited lifetime. The validity period of an access
+token is defined by the server, and may depend on the level of permissions
+granted by the access token, but an hour would be a reasonable
+default.
+
+When a server issues an access token, it may also issue a refresh
+token. Refresh tokens are used to obtain a new access token when the current
+access token expires. This is done via the |/tokenrefresh|_ API. A refresh
```
Comment (ara4n) [inline specification/client_server_api.rst]: The tokenrefresh thing has always bugged me a bit, because:
 * the only protection it provides is slightly reducing the severity when someone leaks an access_token. in practice, i wouldn't want my access_token out in the wild for an hour until it gets cycled though.
 * people can still accidentally leak a refresh_token, it's just a bit more unlikely.
 * it's unproven; no clients or servers implement it to my knowledge.

As such, it feels like it lulls people into a bit of a false sense of safety, and if we want something better we should do some kind of HMAC style thing (and optionally force servers to use HMAC for security), or just piggyback on top of whatever HTTP Auth's best practices are these days.

In other words, whilst `access_token` was a great easy way to bootstrap, rather than glueing something half-assed onto it to make it feel better which doesn't actually buy us much and that we don't actually use, i wonder if we should point people towards a genuinely improved alternative.  And possibly kill the refresh token stuff entirely until that point.

\</thought experiment\>

Comment (richvdh) [inline specification/client_server_api.rst]: so, the access token and refresh token ideas are directly lifted from OAuth2 (https://tools.ietf.org/html/rfc6749#section-1.4 etc). That isn't *necessarily* an argument in its favour, but does mean it's not entirely half-assed.

> the only protection it provides is slightly reducing the severity when someone leaks an access_token.

There are other benefits:
1. If you don't care *too* much about revoking access tokens quickly, it means that you don't need to do a database hit each time you see an access token. For example, in a hypothetical world where stuff on the media server needs auth, it doesn't have to go and ask synapse if an access token is valid on every hit: it just assumes it is until it expires.
2. You can only use a refresh token once, so if an attacker gets hold of one, he blows his cover as soon as he uses it.
3. What if corporate IT policy mandates that the user types in their password every 5 minutes? Admittedly we'd have to spec /tokenrefresh to use the interactive-auth api, but it does provide a mechanism of enforcing that.

(1 is mentioned in the rationale. 2 and 3 possibly should be.)

> As such, it feels like it lulls people into a bit of a false sense of safety, and if we want something better we should do some kind of HMAC style thing 

Well, I think HMAC-signing is slightly orthogonal to whether we use refresh tokens. If you want to keep the three advantages above, you still have to have refresh tokens. It's just a different sort of access token. (Again, see OAuth2: https://tools.ietf.org/html/rfc6749#section-7.1).

On the other hand AWS and co don't seem to bother, so there's a strong argument that it's just needless complication to the spec.

> or just piggyback on top of whatever HTTP Auth's best practices are these days.

Well, I thought that was just OAuth2, at least for RESTful stuff.  Do you know of something else?

There's an argument that we ought to be defining our auth in terms of OAuth2, but the RFC is so damn long that I'd rather just lift the ideas and spec it ourselves.

> And possibly kill the refresh token stuff entirely until that point.

Well, I guess it's adding burden on client implementors which may not be justified. On the other hand, I still think that refresh tokens are a useful thing, and would rather see us make the clients use them, than run in the other direction.

(Also, if we'd decided to can refresh tokens, I kinda wish it had happened *before* I did the device_id stuff...)

File: specification/client_server_api.rst (hunk 1 above)
Comment (richvdh) [inline specification/client_server_api.rst]: probably.

File: specification/client_server_api.rst (hunk 2 above)
Comment (erikjohnston) [inline specification/client_server_api.rst]: > 1. If you don't care too much about revoking access tokens quickly, it means that you don't need to do a database hit each time you see an access token. For example, in a hypothetical world where stuff on the media server needs auth, it doesn't have to go and ask synapse if an access token is valid on every hit: it just assumes it is until it expires.

We still need to check for active revocation though, logging out should really invalidate everything immediately.

> 2 You can only use a refresh token once, so if an attacker gets hold of one, he blows his cover as soon as he uses it.

True. Though maybe there is a better way of doing so. I wonder if we should be able to update the clients token? (Although I suppose that complicates the spec almost as much). The advantage with that is that you can have tokens that expire after N hours of inactivity efficiently (i.e., without hitting the DB)

> 3 What if corporate IT policy mandates that the user types in their password every 5 minutes? Admittedly we'd have to spec /tokenrefresh to use the interactive-auth api, but it does provide a mechanism of enforcing that.

We can auto log people out every 5 minutes without using token refresh? Or do you mean having separate requirements between /login and /tokenrefresh? Adding UIA to /tokenrefresh is really going to complicate things client side.

Comment (richvdh) [inline specification/client_server_api.rst]: https://docs.google.com/document/d/1mdis1LQcoOSVRElszEmrAWZGIX0jX_croSha-X5oe_w/edit#heading=h.bpym87w9ow52 summarises the argument on this. The conclusion seems to be to rip it all out again.

**Comment by richvdh** [comment]
so, as per https://docs.google.com/document/d/1mdis1LQcoOSVRElszEmrAWZGIX0jX_croSha-X5oe_w/edit#heading=h.bpym87w9ow52, we're not doing this.
